### PR TITLE
Remove _pytest.compat.is_generator() fixes #12960

### LINKED ISSUE #12960: Remove `_pytest.compat.is_generator()` and just use `inspect.isgeneratorfunction()` directly
It's already equivalent in practice, and deleting the [few uses and handful of tests](https://github.com/search?q=repo%3Apytest-dev%2Fpytest%20is_generator&type=code) will make it easier for contributors to understand code which uses the stdlib function instead.

While we're there, I'd like to turn the xfail on "yield tests" - a noop since Pytest 4! - into a hard error, to reduce the risk that users incorrectly believe that they're testing something.

### PR DESCRIPTION
**Pull request by arpitgupta-it**
Fixes #12960

### CODE REVIEW

**Comment by arpitgupta-it** [comment]
@Zac-HD I've turned the xfail on 'yield tests' (a noop since Pytest 4) into a hard error, as per the open issue, but it's causing hard failures in the automated pipeline runs for the tests using yield. Do you want me to address these failures as well?

File: changelog/12960.improvement.rst
Comment (Zac-HD) [inline changelog/12960.improvement.rst:1]: This was a private function so we don't need to document it's removal, but we _do_ need to document the `breaking` change to yield tests, e.g.

> Test functions containing a `yield` now cause an explicit error.  They have not been run since Pytest 4.0, and were previously marked as an expected failure and deprecation warning.

File: src/_pytest/python.py
Comment (Zac-HD) [inline src/_pytest/python.py:240]: I think we can simplify this a little further, e.g. 

```python
        elif getattr(obj, "__test__", True):
            if inspect.isgeneratorfunction(obj):
                raise RuntimeError("'yield' keyword is allowed in fixtures, but not in tests ({name})")
            return list(collector._genfunctions(name, obj))
    return None
```

**Review by Zac-HD — COMMENTED** [review]
Thanks, @arpitgupta-it!

We'll need to get all the tests passing before we merge - that's how we can tell that the PR didn't break anything unexpected, after all.  Based on [this case](https://github.com/pytest-dev/pytest/blob/71a35d4a3c8bf626cb73be0cb900ede18b1b123d/testing/test_terminal.py#L1045-L1048) I'd guess that most of the failing tests can be fixed just by deleting the now-an-error `yield`-test, and any associated [xfail expectations](https://github.com/pytest-dev/pytest/blob/71a35d4a3c8bf626cb73be0cb900ede18b1b123d/testing/test_terminal.py#L1061).

**Review by Zac-HD — COMMENTED** [review]
One quick fix for the changelog entry, and then I think we're ready to merge!

cc @RonnyPfannschmidt, I'll merge in a couple of days unless you object.

File: changelog/12960.improvement.rst
Comment (arpitgupta-it) [reply changelog/12960.improvement.rst:1]: Done!

**Review by Zac-HD — APPROVED** [review]

File: changelog/12960.breaking.rst
Comment (nicoddemus) [inline changelog/12960.breaking.rst:1]: There's a section named `yield tests` in `deprecations.rst`, under `Deprecated Features`. Given this is actually being removed, please move that section to `Removed Features and Breaking Changes`, adding a `.. versionremoved:: 8.4` to it (see the others for an example).

Comment (nicoddemus) [inline changelog/12960.breaking.rst:1]:
```suggestion
Test functions containing a yield now cause an explicit error. They have not been run since pytest 4.0, and were previously marked as an expected failure and deprecation warning.

See :ref:`the docs <yield tests deprecated>` for more information.
```

File: src/_pytest/python.py
Comment (nicoddemus) [inline src/_pytest/python.py:234]: An internal error will unfortunately show an internal traceback which is quite noisy. Let's use `fail` instead (also it was missing the `f` for the f-string):

```suggestion
                fail(
                    f"'yield' keyword is allowed in fixtures, but not in tests ({name})",
                    pytrace=False,
```

File: testing/test_collection.py
Comment (nicoddemus) [inline testing/test_collection.py:1894]: Please make sure to match the test name in the message (there ia  bug in the original code which is not caught because of that).

File: testing/test_collection.py
Comment (nicoddemus) [inline testing/test_collection.py:1884]:
```suggestion
    """Ensure generator test functions with 'yield' fail colletion (#12960)."""
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @arpitgupta-it! Please take a look at my comments.

**Comment by arpitgupta-it** [comment]
@nicoddemus @Zac-HD @RonnyPfannschmidt are we good to go?

File: changelog/12960.breaking.rst
Comment (nicoddemus) [inline changelog/12960.breaking.rst:1]:
```suggestion
Test functions containing a yield now cause an explicit error. They have not been run since pytest 4.0, and were previously marked as an expected failure and deprecation warning.

See :ref:`the docs <yield tests deprecated>` for more information.
```

**Review by nicoddemus — APPROVED** [review]

**Comment by nicoddemus** [comment]
Thanks @arpitgupta-it!

I will leave it open for a few more days in case others want to review it, otherwise I will merge.

File: doc/en/deprecations.rst
Comment (Zac-HD) [inline doc/en/deprecations.rst:382]: Ah, I think this is actually wrong - they were removed long ago, but in a way where the function was marked as xfail without running it; we just made it an error rather than an xfail now.  I don't know exactly when the transition was but I worry that this doc snippet might make users think that yield tests worked on 8.3, for example.

Comment (nicoddemus) [reply doc/en/deprecations.rst:382]: Hmm good point, how about:

```suggestion
.. versionremoved:: 4.0

    ``yield`` tests ``xfail``.

.. versionremoved:: 8.4

    ``yield`` tests raise a collection error.
```

This is rendered like this:

image

Comment (Zac-HD) [reply doc/en/deprecations.rst:382]: That looks great.

**Comment by arpitgupta-it** [comment]
Sounds like a good plan—I'll keep an eye out in case anyone else has feedback. Let me know if you guys need anything else from my side before the merge!

**Comment by Zac-HD** [comment]
Okay, I think with two core devs happy it's time to merge!  Thanks again @arpitgupta-it 🙏

**Comment by nicoddemus** [comment]
@Zac-HD thanks! However in the future please try to rebase/squash commits, as unfortunately a lot of "updated changelog", merge, etc commits landed on main.

Please see https://github.com/pytest-dev/pytest/blob/main/CONTRIBUTING.rst#mergesquash-guidelines in case you missed it. 👍

**Pull request merged**
